ref(breadcrumbs): Improve crumbs docs (#2067)
getsentry/sentry-docs: src/docs/enriching-error-data/breadcrumbs/index.mdx

---
title: Breadcrumbs
sidebar_order: 2
redirect_from:
  - /learn/breadcrumbs/
---

Sentry supports a concept called Breadcrumbs, which is a trail of events that happened prior to an issue. Oftentimes these events are very similar to traditional logs, but also have the ability to record more rich structured data.

<Alert level="" title="Learn about SDK usage"><markdown>

> This page will provide an overview of of the Breadcrumbs section. Developers who want to modify the breadcrumbs interface can read about this in detail using the developer documentation devoted to the [Breadcrumbs Interface](https://develop.sentry.dev/sdk/event-payloads/breadcrumbs/). Manual breadcrumb recording and customization of breadcrumbs are documented on this page.

Reviewer: ```suggestion
This page will provide an overview of the Breadcrumbs section. Developers who want to modify the breadcrumbs interface can read about this in detail using the developer documentation devoted to the [Breadcrumbs Interface](https://develop.sentry.dev/sdk/event-payloads/breadcrumbs/). Manual breadcrumb recording and customization of breadcrumbs are documented on this page.
```

</markdown></Alert>

If an event of type error provides a trail, Sentry will display that trail in a form of breadcrumbs on the event’s issues details page. It looks like:

![Breadcrumb section](crumb-section.png)

## Table

The crumb’s table renders each trail a separate row and it’s information in 5 different columns: `type`, `category`, `description`, `level`, and `time`.

> _You can read more about the columns in a crumb document under a platform docs._

Reviewer: ```suggestion
_You can read more about the columns in a crumb document under platform docs._
```

The column `time` displays alongside it's title a toggle button, providing a way to switch from absolute time format to relative time format.

![Switch time format button](toggle-time-btn.png)

By default when the table’s renders, it will automatically scroll to the `exception/error` at the bottom.

## Filter

#### Filter By

The provided types and levels are displayed as options in this filter, as for example:

![Filter By Types/Levels](filter-by.png)

The crumb’s list will be filtered according to the selected options.

#### Full-Text Search

When typing anything in the input field, Sentry filters out all fragmented data, trying to find a match to the search criteria. If a match is found, it will be highlighted.

_Be aware that the full-text search works in conjunction with the ‘Filter By’ filter._

> ## Recording Crumbs
> 
> Manual breadcrumb recording is also available and easy to use. This way breadcrumbs
> can be added whenever something interesting happens. For instance, it might make sense
> to record a breadcrumb if the user authenticates or another state change happens.
> 
> <PlatformContent includePath="breadcrumbs-example" />
> 
> ## Automatic Breadcrumbs
> 
> SDKs will automatically start recording breadcrumbs by enabling integrations. For instance,
> the browser JavaScript SDK will automatically record all location changes.
> 
> ## Breadcrumb Customization
> 
> SDKs following the unified API support customizing breadcrumbs through the `before_breadcrumb`
> hook. This hook is passed an already assembled breadcrumb and in some SDKs an optional
> hint. The function can modify the breadcrumb or decide to discard it entirely:
> 
> <PlatformContent includePath="before-breadcrumb" />

Reviewer: I would remove this section and guide users towards to the platforms / sdk docs how to do this there.
```suggestion
## Recording Crumbs

Manual breadcrumb recording is also available and easy to use. This way breadcrumbs
can be added whenever something interesting happens. For instance, it might make sense
to record a breadcrumb if the user authenticates or another state change happens.

To figure out how to do this for your platform, select it in the sidebar and dive into the breadcrumbs section.
```

cc @PeloWriter The Platform selection page is gone, maybe we update this here later to guide people there.

Author: I wanted to do this, but I decided to leave it as it is until we have a breadcrumb doc under each platform. Please @PeloWriter what do you think?

Reviewer 2: Yes, I agree. Until we have the new format adopted by the SDKs, this is a good interim approach.

Reviewer 2: The platform/SDK content in JS (our model) and the general page now have the same content, so it will be easy to move away from the general page once the SDKs are migrated to the new format. I think this work as an interim solutionL

SDKs automatically start recording breadcrumbs by enabling integrations. To learn how to do this for your platform, select the platform from the sidebar, then dive into the breadcrumbs documentation for that platform

Author: love it! Thank you so much Fiona 🙏
